### PR TITLE
feat: Add support for cloudwatch cross account oberservability #12447

### LINKED ISSUE #12447: Add support for cloudwatch cross account oberservability
### Use Case

At the moment, if we want to collect metrics from cloudwatch and store them in influxDB, we must run telegraf in each account, which is inefficient and expensive in terms of getdatametricdata api calls. We can use cross account observability to designate a single account as a monitoring account and then pull metrics from multiple accounts, eliminating the operational overhead of managing multiple telegrafs and significantly lowering the cost of API calls.

### Expected behavior

Run a telegraf agent with cloudwatch input in each AWS account 

### Actual behavior

Run a telegraf agent in a designated monitoring account and pull metrics from all/multiple accounts with in an org 

### Additional info

https://aws.amazon.com/blogs/aws/new-amazon-cloudwatch-cross-account-observability/

### PR DESCRIPTION
**Pull request by ahsanali**
resolves #12447

### CODE REVIEW

**Comment by ahsanali** [comment]
!signed-cla

**Comment by ahsanali** [comment]
thanks for the review @powersj

**Review by powersj — APPROVED** [review]

File: plugins/inputs/cloudwatch/README.md [hunk 1]
```diff
@@ -226,6 +231,8 @@ case](https://en.wikipedia.org/wiki/Snake_case)
 - All measurements have the following tags:
   - region           (CloudWatch Region)
   - {dimension-name} (Cloudwatch Dimension value - one per metric dimension)
+- If IncludeLinkedAccounts is set to true then below tag is also provided:
+  - account           (The ID of the account where the metrics are located.)
```
Comment (srebhan) [inline plugins/inputs/cloudwatch/README.md:235]:
```suggestion
- If `include_linked_accounts` is set to true then below tag is also provided:
  - account           (The ID of the account where the metrics are located.)
```

File: plugins/inputs/cloudwatch/sample.conf [hunk 2]
```diff
@@ -21,7 +21,12 @@
   # role_session_name = ""
   # profile = ""
   # shared_credential_file = ""
-
+  
+  ## If you are using CloudWatch cross-account observability, you can 
+  ## set IncludeLinkedAccounts to true in a monitoring account 
+  ## and collect metrics from the linked source accounts
+  # IncludeLinkedAccounts = false
+
```
Comment (srebhan) [inline plugins/inputs/cloudwatch/sample.conf:29]: Please use snake_case for options! I think you also did for the TOML annotation in code, but did not yet for the documentation
```suggestion
  ## If you are using CloudWatch cross-account observability, you can 
  ## set IncludeLinkedAccounts to true in a monitoring account 
  ## and collect metrics from the linked source accounts
  # include_linked_accounts = false
  
```

File: plugins/inputs/cloudwatch/cloudwatch.go [hunk 3]
```diff
@@ -367,66 +366,37 @@ func (c *CloudWatch) getDataQueries(filteredMetrics []filteredMetric) map[string
...
+						Id:        aws.String(queryID),
+						AccountId: accountID,
+						Label:     aws.String(snakeCase(*metric.MetricName + "_" + statisticType)),
+						MetricStat: &types.MetricStat{
+							Metric: &filtered.metrics[j],
+							Period: aws.Int32(int32(time.Duration(c.Period).Seconds())),
+							Stat:   aws.String(statistic),
+						},
+					})
+				}
```
Comment (srebhan) [inline plugins/inputs/cloudwatch/cloudwatch.go:397]: Can we please save one level of indentation via
```suggestion
				if !filtered.statFilter.Match(statisticType) {
					continue
				}
				queryID := statisticType + "_" + id
				c.queryDimensions[queryID] = dimension
				dataQueries[*metric.Namespace] = append(dataQueries[*metric.Namespace], types.MetricDataQuery{
					Id:        aws.String(queryID),
					AccountId: accountID,
					Label:     aws.String(snakeCase(*metric.MetricName + "_" + statisticType)),
					MetricStat: &types.MetricStat{
						Metric: &filtered.metrics[j],
						Period: aws.Int32(int32(time.Duration(c.Period).Seconds())),
						Stat:   aws.String(statistic),
					},
				})
```

**Review by srebhan — COMMENTED** [review]
Thanks for this nice feature @ahsanali! I found an inconsistency in your sample config and have two small requests for code changes. Nothing too big. Can you please take a look?

File: plugins/inputs/cloudwatch/cloudwatch.go (hunk 3 above)
Comment (ahsanali) [reply plugins/inputs/cloudwatch/cloudwatch.go:397]: feat: adding support for cloudwatch cross account observability. reindenting the code

**Comment by powersj** [comment]
@ahsanali would be happy to see this updated so we can land it. Just let us know in a comment here

**Comment by ahsanali** [comment]
@powersj sorry for the delay. I have updated the fork. Should I create a new PR ?

**Comment by powersj** [comment]
@ahsanali I have re-opened the PR. Could you push your additional changes and ping @srebhan when you are ready for another review?

Thanks!

**Comment by khurramcheema** [comment]
@srebhan please review and merge it, Thanks

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for the nice feature @ahsanali!

**Pull request closed without merging**